src/mito2/src/sst/index/puffin_manager.rs

```diff
@@ -0,0 +1,207 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::Path;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use common_error::ext::BoxedError;
+use object_store::{FuturesAsyncReader, FuturesAsyncWriter, ObjectStore};
+use puffin::error::{self as puffin_error, Result as PuffinResult};
+use puffin::puffin_manager::file_accessor::PuffinFileAccessor;
+use puffin::puffin_manager::fs_puffin_manager::FsPuffinManager;
+use puffin::puffin_manager::stager::{BoundedStager, FsBlobGuard, FsDirGuard};
+use puffin::puffin_manager::BlobGuard;
+use snafu::ResultExt;
+
+use crate::error::{PuffinInitStagerSnafu, Result};
+use crate::metrics::{
+    INDEX_PUFFIN_FLUSH_OP_TOTAL, INDEX_PUFFIN_READ_BYTES_TOTAL, INDEX_PUFFIN_READ_OP_TOTAL,
+    INDEX_PUFFIN_SEEK_OP_TOTAL, INDEX_PUFFIN_WRITE_BYTES_TOTAL, INDEX_PUFFIN_WRITE_OP_TOTAL,
+};
+use crate::sst::index::store::{self, InstrumentedStore};
+
+type InstrumentedAsyncRead = store::InstrumentedAsyncRead<'static, FuturesAsyncReader>;
+type InstrumentedAsyncWrite = store::InstrumentedAsyncWrite<'static, FuturesAsyncWriter>;
+
+pub(crate) type BlobReader = <Arc<FsBlobGuard> as BlobGuard>::Reader;
+pub(crate) type SstPuffinManager = FsPuffinManager<
+    Arc<FsBlobGuard>,
+    Arc<FsDirGuard>,
+    InstrumentedAsyncRead,
+    InstrumentedAsyncWrite,
+>;
+
+const STAGING_DIR: &str = "staging";
+
+/// A factory for creating `SstPuffinManager` instances.
+#[derive(Clone)]
+pub struct PuffinManagerFactory {
+    /// The stager used by the puffin manager.
+    stager: Arc<BoundedStager>,
+
+    /// The size of the write buffer used to create object store.
+    write_buffer_size: Option<usize>,
+}
+
+impl PuffinManagerFactory {
+    /// Creates a new `PuffinManagerFactory` instance.
+    pub async fn new(
+        auxiliary_path: impl AsRef<Path>,
+        staging_capacity: u64,
+        write_buffer_size: Option<usize>,
+    ) -> Result<Self> {
+        let staging_dir = auxiliary_path.as_ref().join(STAGING_DIR);
+        let stager = BoundedStager::new(staging_dir, staging_capacity)
+            .await
+            .context(PuffinInitStagerSnafu)?;
+        Ok(Self {
+            stager: Arc::new(stager),
+            write_buffer_size,
+        })
+    }
+
+    pub(crate) fn build(&self, store: ObjectStore) -> SstPuffinManager {
+        let store = InstrumentedStore::new(store).with_write_buffer_size(self.write_buffer_size);
+        let puffin_file_accessor = ObjectStorePuffinFileAccessor::new(store);
+        SstPuffinManager::new(self.stager.clone(), Arc::new(puffin_file_accessor))
```

Comment:
Maybe we should refactor the puffin manager to take a stager and a file accessor as generic parameters
```rust
pub struct FsPuffinManager<S, F> {
    /// The stager.
    stager: S,

    /// The puffin file accessor.
    puffin_file_accessor: F,
}

impl<S: Stager, F: FileAccessor> PuffinManager for FsPuffinManager<S, F> {}
```

It reduces the number of generic parameters (from 4 to 2) and allows us to pass `ObjectStorePuffinFileAccessor` directly without allocating an `Arc` each time.

But we may do this in a new PR.